addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,588 @@
+	key := client.ObjectKey{Namespace: constants.TKGSystemNS, Name: tkrName}
+	if err := r.Client.Get(r.context, key, tanzuClusterBootstrapTemplate); err != nil {
+		log.Error(err, "unable to fetch tanzuClusterBootstrapTemplate", "objectkey", key)
+		return nil, err
+	}
+
+	tanzuClusterBootstrap := &runtanzuv1alpha3.TanzuClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), tanzuClusterBootstrap)
+	// if found and resolved tkr is the same, return found object as the TKR is supposed to be immutable
+	// also preserves any user changes
+	if err == nil && tkrName == tanzuClusterBootstrap.Status.ResolvedTKR {
+		return tanzuClusterBootstrap, nil
+	}
+
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if tanzuClusterBootstrap.UID == "" {
+		log.Info("tanzuClusterBootstrap for cluster does not exist, cloning from template")
+
+		tanzuClusterBootstrap.Name = cluster.Name
+		tanzuClusterBootstrap.Namespace = cluster.Namespace
+		tanzuClusterBootstrap.Spec = tanzuClusterBootstrapTemplate.Spec.DeepCopy()
+		secrets, providers, err := r.cloneSecretsAndProviders(cluster, tanzuClusterBootstrap, tanzuClusterBootstrapTemplate.Namespace, log)
+		if err != nil {
+			r.Log.Error(err, "unable to clone secrets, providers")
+			return nil, err
+		}
+
+		tanzuClusterBootstrap.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+				Kind:               cluster.Kind,
+				Name:               cluster.Name,
+				UID:                cluster.UID,
+				Controller:         pointer.BoolPtr(true),
+				BlockOwnerDeletion: pointer.BoolPtr(true),
+			},
+		}
+
+		if err := r.Client.Create(r.context, tanzuClusterBootstrap); err != nil {
+			return nil, err
+		}
+		// ensure ownerRef of tanzuClusterBootstrap on created secrets and providers, this can only be done after
+		// tanzuClusterBootstrap is created
+		ownerRef := metav1.OwnerReference{
+			APIVersion:         runtanzuv1alpha3.GroupVersion.String(),
+			Kind:               "TanzuClusterBootstrap", // kind is empty after create
+			Name:               tanzuClusterBootstrap.Name,
+			UID:                tanzuClusterBootstrap.UID,
+			Controller:         pointer.BoolPtr(true),
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}
+		if err := r.ensureOwnerRef(&ownerRef, secrets, providers); err != nil {
+			r.Log.Error(err, "unable to ensure ownerref on created secrets and providers", "tanzuclusterbootstrap", tanzuClusterBootstrap)
+			return nil, err
+		}
+
+		tanzuClusterBootstrap.Status.ResolvedTKR = tkrName
+		if err := r.Status().Update(r.context, tanzuClusterBootstrap); err != nil {
+			return nil, err
+		}
+		r.Log.Info("cloned tanzuclusterbootstrap", "tanzuclusterbootstrap", tanzuClusterBootstrap)
+		return tanzuClusterBootstrap, nil
+	}
+
+	// TODO upgrade needs patch (update versions of all packages, merge configs, add additional packages and remove packages that don't exist anymore)
```

Comment:
suggestion: Can we add/find an issue and refer to it here and in the PR?

Please disregard if this is still WIP.

Reply:
Same as earlier, see https://github.com/orgs/vmware-tanzu/projects/15/views/22. We discussed when reviewing within team to add TODOs to make it easier for other people to pickup.

Reply:
I'm all for TODOs when we know something needs to be done! I just want these things to be surfaced as issues. I also don't want them to be forgotten. And we, as humans, can forget about them in a couple different ways:

1. Forget what the TODO means (multiple examples to this, right in this repo)
2. Forget TO DO it: because there's no issue, and there's a lot of other issues to address, we address them and forget about the TODO which often leads to (1).